### PR TITLE
Add mutable fields

### LINKED ISSUE #8: Add mutable assignment
The scheme is to use an invariant `Ref` type (or call it `Mut`) everywhere, where we have `Ref[T] <: T`, and with a `Ref: 'a -> Ref['a]` constructor. For example we have `{ x = Ref 1 } : { x: Ref[0 | 1] } <: { x: int }`

The advantage over explicitly making record fields optionally mutable is to make the implementation in MLscript easy and modular. But we need to put a restriction on where these `Ref` types can appear. The restriction could be checked in a pass after type checking:
**_One can only call the `Ref` constructor as part of a field initialization, as in `{ x = Ref 1 }` or as part of a local variable initialization, as in `let x = Ref 1 in ...`._**

The code generator will simply ignore these Ref constructors, as in JS all fields are mutable by default.

### PR DESCRIPTION
**Pull request by LPTK**
Closes #8.

### CODE REVIEW

File: shared/src/main/scala/mlscript/Typer.scala [hunk 1]
```diff
@@ -414,6 +414,7 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
                           val fv = freshVar(noProv,
                             S(f._1.name.drop(f._1.name.indexOf('#') + 1)) // strip any "...#" prefix
                           )(1).tap(_.upperBounds ::= f._2.ub)
+                          // ? not sure if we can do this
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/Typer.scala]: This is a bit funky. It may work in most cases but that's not really the logic we want. Here, we should decide based on whether the field was declared mutable or not. I think you should do what I suggested on Discord:

> it may be better to make the LB of FieldType an Option, and if it's None it means it's immutable

Then things will be a lot cleaner, including here.

This can also give better errors.

Comment (Meowcolm024) [reply shared/src/main/scala/mlscript/Typer.scala]: So it means that we're using `None` for place of `BotType` originally(for immutable ones), and use `Some(T)` for the mutable ones? @_@

Comment (LPTK) [reply shared/src/main/scala/mlscript/Typer.scala]: Yes, exactly!

Comment (Meowcolm024) [reply shared/src/main/scala/mlscript/Typer.scala]: But it seems that the lower bound would always equals to the upper bound for mutable fields, so what is the difference between this way and simply using a flag @_@
Or is there scenario where the lower bound is different?

Comment (LPTK) [reply shared/src/main/scala/mlscript/Typer.scala]: The problem is that if we used a single bound, it would be invariant, which would make other parts of the compiler awkward because they currently assume that everything is either covariant or contravariant. For example, merging two mutable tuples in type simplification would require making the two types equal by constraining, but currently no constraining occurs in simplification. It's more of a technical detail, really.

File: shared/src/main/scala/mlscript/TypeSimplifier.scala [hunk 2]
```diff
@@ -385,13 +388,14 @@ trait TypeSimplifier { self: Typer =>
...
                       val componentFieldsMap = componentFields.toMap
                       val tupleComponents = fs.iterator.zipWithIndex.map { case ((nme, ty), i) =>
-                        nme -> go(ty & componentFieldsMap.getOrElse(i + 1, TopType), pol)
+                        nme -> (ty && componentFieldsMap.getOrElse(i + 1, TopType.toUpper(noProv))).update(go(_, !pol), go(_, pol))
                       }.toList
-                      S(TupleType(tupleComponents)(tt.prov)) -> rcdFields.mapValues(go(_, pol))
+                      S(TupleType(tupleComponents)(tt.prov)) -> rcdFields.mapValues(_.update(go(_, !pol), go(_, pol)))
                     case S(ct: ClassTag) => S(ct) -> nFields
-                    case S(ft @ FunctionType(l, r)) =>
+                    case S(ft @ FunctionType(l, r)) =>
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/TypeSimplifier.scala]:
```suggestion
                    case S(ft @ FunctionType(l, r)) =>
```

File: shared/src/main/scala/mlscript/TyperDatatypes.scala [hunk 3]
```diff
@@ -253,12 +254,38 @@ abstract class TyperDatatypes extends TyperHelpers { self: Typer =>
   case class TypeBounds(lb: SimpleType, ub: SimpleType)(val prov: TypeProvenance) extends SimpleType {
     def level: Int = lb.level max ub.level
     override def toString = s"$lb..$ub"
+    def && (that: TypeBounds, prov: TypeProvenance = noProv): TypeBounds =
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/TyperDatatypes.scala]: Question: is this actually still used?

File: shared/src/main/scala/mlscript/TypeSimplifier.scala (hunk 2 above)
Comment (LPTK) [reply shared/src/main/scala/mlscript/TypeSimplifier.scala]: IS this resolved? If so, please mark it so.

Comment (Meowcolm024) [reply shared/src/main/scala/mlscript/TypeSimplifier.scala]: yep

Comment (LPTK) [reply shared/src/main/scala/mlscript/TypeSimplifier.scala]: > If so, please mark it so. (resolved)

File: shared/src/main/scala/mlscript/JSBackend.scala [hunk 4]
```diff
@@ -203,6 +204,7 @@ class JSBackend {
       JSArray(terms map { case (_, (term, _)) => translateTerm(term) })
     case Subs(arr, idx) =>
       JSMember(translateTerm(arr), translateTerm(idx))
+    case Assign(field, value) => ???  // TODO
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/JSBackend.scala]: I think it should be easy to implement this case while we're at it. You just need to generate a JS assignment expression.

File: shared/src/main/scala/mlscript/TyperDatatypes.scala (hunk 3 above)
Comment (Meowcolm024) [reply shared/src/main/scala/mlscript/TyperDatatypes.scala]: Seems not used from code lens, both`&&` and `||`

Comment (LPTK) [reply shared/src/main/scala/mlscript/TyperDatatypes.scala]: Okay, I think we can remove it then.

File: shared/src/main/scala/mlscript/JSBackend.scala (hunk 4 above)
Comment (Meowcolm024) [reply shared/src/main/scala/mlscript/JSBackend.scala]: I find that I also need to return a `unit`, but I don't know how @_@

Comment (Meowcolm024) [reply shared/src/main/scala/mlscript/JSBackend.scala]: Okay, I think I've find a way

Comment (LPTK) [reply shared/src/main/scala/mlscript/JSBackend.scala]: You can return an empty array as unit, which is the current interpretation:

```scala
()
//│ res: ()
//│    = []
```

File: shared/src/main/scala/mlscript/JSBackend.scala [hunk 5]
```diff
@@ -51,7 +51,7 @@ class JSBackend {
     case Bra(_, trm) => translatePattern(trm)
     case Tup(fields) => JSArrayPattern(fields map { case (_, (t, _)) => translatePattern(t) })
     // Others are not supported yet.
-    case _: Lam | _: App | _: Sel | _: Let | _: Blk | _: Bind
+    case _: Lam  | _: App  | _: Sel    | _: Let  | _: Blk    | _: Bind
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/JSBackend.scala]: Please don't... Horizontal alignment is a sin except in a few exceptional cases (such as aligning `case ... =>`).

File: shared/src/main/scala/mlscript/JSBackend.scala [hunk 6]
```diff
@@ -204,11 +204,8 @@ class JSBackend {
       JSArray(terms map { case (_, (term, _)) => translateTerm(term) })
     case Subs(arr, idx) =>
       JSMember(translateTerm(arr), translateTerm(idx))
-    case Assign(field, value) => JSImmEvalFn(N, Nil, Left(
-      JSCommaExpr(
-        JSAssignExpr(translateTerm(field), translateTerm(value)) :: 
-        JSArray(Nil) :: Nil   // return unit
-      )), Nil)
+    case Assign(field, value) => JSImmEvalFn(N, JSNamePattern("_") :: Nil, Left(JSArray(Nil)), 
+      JSAssignExpr(translateTerm(field), translateTerm(value)) :: Nil)
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/JSBackend.scala]: So why not use a comma expression here? As in `(x = a, [])`.

File: shared/src/main/scala/mlscript/TypeDefs.scala [hunk 7]
```diff
@@ -133,7 +133,7 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
   
   
   def processTypeDefs(newDefs0: List[mlscript.TypeDef])(implicit ctx: Ctx, raise: Raise): Ctx = {
-    
+
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/TypeDefs.scala]: Can you please _not_ remove the indentation space? Please revert all these indentation deletions, they're unnecessary changes.

**Review by LPTK — COMMENTED** [review]
Nice, LGTM. I thin it's almost ready to merge except for my three comments below.

File: shared/src/main/scala/mlscript/JSBackend.scala (hunk 6 above)
Comment (Meowcolm024) [reply shared/src/main/scala/mlscript/JSBackend.scala]: I'm not sure if using a comma is suitable @_@, I wonder if there is a semicolon that I can use

Comment (LPTK) [reply shared/src/main/scala/mlscript/JSBackend.scala]: AFAIK that's how JS does sequencing of effectful expressions in an expression. JS does not have first-class `;` nor expression blocks like Scala.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator

File: shared/src/main/scala/mlscript/JSBackend.scala [hunk 8]
```diff
@@ -194,15 +194,18 @@ class JSBackend {
...
           name -> translateTerm(value)
         }) :: Nil
       )
     case Bra(_, trm) => translateTerm(trm)
     case Tup(terms) =>
-      JSArray(terms map { case (_, term) => translateTerm(term) })
+      JSArray(terms map { case (_, (term, _)) => translateTerm(term) })
     case Subs(arr, idx) =>
       JSMember(translateTerm(arr), translateTerm(idx))
+    case Assign(field, value) => JSImmEvalFn(N, Nil,
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/JSBackend.scala]: But what is the point of wrapping thins in a function?

Comment (Meowcolm024) [reply shared/src/main/scala/mlscript/JSBackend.scala]: It seems that there's no need @_@

**Comment by LPTK** [comment]
Okay, merging now!
